launcher/launcher.go

```diff
@@ -101,28 +108,42 @@ func (l *launcher) Run(b *beat.Beat) error {
 }
 
 func (l *launcher) run() error {
-	l.log.Info("Beater launcher is running")
+	l.log.Info("Launcher is running")
 	err := l.runBeater()
 	if err != nil {
-		l.log.Errorf("Could not run Beater: %v", err)
+		l.log.Errorf("Launcher could not run Beater: %v", err)
 		return err
 	}
 
-	err = l.waitForUpdates()
+	go l.waitForUpdates()
+	err = l.waitForFinish()
 	if err != nil {
-		l.log.Errorf("Beater launcher is stopping: %v", err)
+		l.log.Errorf("Launcher has stopped: %v", err)
 	} else {
-		l.log.Info("Beater launcher was shutted down gracefully")
+		l.log.Info("Launcher was shutted down gracefully")
 	}
 	return err
 }
 
 func (l *launcher) Stop() {
-	l.log.Info("Beater launcher is about to shut down gracefully")
+	l.log.Info("Launcher is about to shut down gracefully")
+
+	// Stop listening for updates
+	l.reloader.Stop()
+
+	// Stop the beater without interrupting to an update
+	l.cfgmtx.Lock()
+	l.runmtx.Lock()
```

Comment:
I wonder if `defer l.mtx.Unlock()` is relevant also here?

Will the next run happen with a new launcher for sure?